website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,196 @@
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 34                 | 1.0                     |
+| Baseline | 18                  | 253                | 7.44                    |
+
+Now, instead of having a filter that is always true, we retrieve the value of `extendedprice` and materialize a struct. This takes only marginally longer because the struct only has the `extendedprice` column filled in. The struct needs to be materialized because it is an input to the SUM aggregate. We could of course, since we only access fields of the struct and not the whole struct, elide materializing the struct and only materialize the component columns. But the gain of this last optimization would not be very large.
+
+```sql
+SELECT COUNT(*), SUM(shipment.extendedprice), COUNT(export.license.filing_no)
+FROM exportlineitem
+WHERE shipment.suppkey BETWEEN 200000 AND 400000 AND shipment.quantity < 10;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 10                  | 58                 | 1.0                     |
+| Baseline | 30                  | 330                | 5.68                    |
+
+Here we add a second struct to the mix. We filter on members of one struct and return a field, `filing_no`, which is wrapped inside two structs. Both the `export` struct and the `license` struct inside it are nullable, i.e. not all shipments are international and not all export shipments need a license.
+This takes longer because we must read two levels of null flags, one for `export` and the other for the `license` substruct. Then we read the filing number for the positions where there is a `license` and fill in a null for the case where either `license` or the enclosing `export` struct is null.
+
+# Experiments
+
+```sql
+SELECT SUM(shipment.extendedprice), COUNT(export.license.filing_no)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 200000 AND 400000 AND shipment.quantity < 10 AND export.s_nation IN (1, 3, 6) AND export.c_nation = 11;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 5                   | 50                 | 1.0                     |
+| Baseline | 30                  | 309                | 5.32                    |
+
+Here we add a filter on `export`. The filters within structs are reorderable as well as the top level structs. We find that the new filter does not negatively impact running time and in fact can improve it. The best filter is evaluated first and after this all column access is sparse and only the data at interesting positions gets touched.
```

Comment:
Another surprise. CPU improved a little bit - 8%, but wall time dropped in half. Do we have an explanation? Is this just the variability of the experiment? If so, it seems like we should run the same multiple times and take the average. Otherwise, the numbers appear arbitrary.

Reply:
The wall time is what's off, we re-ran and it's ~5 second every time. This query is more selective with an additional 3/25 and 1/25 filter and with reordering the most efficient of those will run first.

Wall time is generally a bit more noisy than CPU. I also find that wall times vary the most when things aren't warmed up.
```
CPU
51
51.6
50.2
49.9
49.2
48.8
48.6
49.2

WALL
5.995
5.913
5.845
5.1
4.904
5.883
```

Reply:
Pushed an update with an average of the new measurements. In future blog posts with performance numbers I'll ask Orri to provide at least 3 measurements for each query.